src/core/thread/mesh_forwarder.cpp

```diff
@@ -1840,7 +1840,7 @@ uint16_t MeshForwarder::CalcFrameVersion(const Neighbor *aNeighbor, bool aIePres
     }
 #if OPENTHREAD_FTD && OPENTHREAD_CONFIG_MAC_CSL_TRANSMITTER_ENABLE
     else if (aNeighbor != nullptr && !Mle::MleRouter::IsActiveRouter(aNeighbor->GetRloc16()) &&
-             static_cast<const Child *>(aNeighbor)->IsCslSynchronized())
+             Get<Mle::MleRouter>().IsRouterOrLeader() && static_cast<const Child *>(aNeighbor)->IsCslSynchronized())
```

Comment:
Thanks @canisLupus1313 for finding this.  👍 
Indeed a tricky bug. I guess this can be caused by a `Parent` (or `ParentCandidate`) which are still acting as a REED.

I am thinking if it may be safer for us to use `ChildTabele::Contains(Neighbor &)` method here before casting to `Child *`. This will check that the `Neighbor` is from child table and the cast is safe/correct. I think the `IsRouterOrleader()` will indirectly check the same thing but I feel direct check may be better.